lib/spack/spack/database.py

```diff
@@ -1648,32 +1658,49 @@ def is_occupied_install_prefix(self, path):
         with self.read_transaction():
             return path in self._installed_prefixes
 
-    @property
-    def unused_specs(self):
-        """Return all the specs that are currently installed but not needed
-        at runtime to satisfy user's requests.
-
-        Specs in the return list are those which are not either:
-            1. Installed on an explicit user request
-            2. Installed as a "run" or "link" dependency (even transitive) of
-               a spec at point 1.
+    def all_hashes(self):
+        """Return dag hash of every spec in the database."""
+        return list(self._data.keys())
+
+    def unused_specs(
+        self,
+        root_hashes: Optional[Container[str]] = None,
+        deptypes: Optional[Tuple[str, ...]] = None,
+    ) -> "List[spack.spec.Spec]":
+        """Return all specs that are currently installed but not needed by root specs.
+
+        By default, roots are all explicit specs in the database. If a set of root
+        hashes are passed in, they are instead used as the roots.
+
+        Arguments:
+            root_hashes: optional list of roots to consider when evaluating needed installations.
+            deptypes: if a spec is reachable from a root via these dependency types, it is
+                considered needed. By default only link and run dependency types are considered.
         """
-        needed, visited = set(), set()
+        deptypes = ("link", "run") if deptypes is None else deptypes
+
+        def root(key, record):
+            """Whether a DB record is a root for garbage collection."""
+            return key in root_hashes if root_hashes is not None else record.explicit
+
+        needed: Set[str] = set()
+        visited: Set[spack.spec.Spec] = set()
+
         with self.read_transaction():
-            for key, rec in self._data.items():
-                if not rec.explicit:
+            for key, record in self._data.items():
```

Comment:
Maybe simpler is to collect `roots = [s for s in ... if root(s)]` and then use `spack.traverse.traverse_nodes(roots)`, and drop the `visited` set.

Reply:
Yeah this simplified it quite a bit. I like it.

I'm not sure how I feel about `traverse_nodes()` being a generator. It was somewhat annoying to realize you can't just pass it the visited set and execute it for side effects -- you have to actually consume the generator or invoking `traverse_nodes()` does nothing. I suppose that is expected.

Luckily you can fairly easily make a one-liner for this, so it turned out ok, but it feels awkward to make a second set when you could just pass one in.